src/tail/session.rs

```diff
@@ -0,0 +1,118 @@
+use std::str;
+use std::thread;
+use std::time::Duration;
+
+use regex::Regex;
+use reqwest;
+use tokio::sync::oneshot::error::TryRecvError;
+use tokio::sync::oneshot::Receiver;
+use tokio::time;
+
+use cloudflare::endpoints::workers::{CreateTail, CreateTailParams, SendTailHeartbeat};
+use cloudflare::framework::HttpApiClientConfig;
+use cloudflare::framework::{async_api, async_api::ApiClient};
+
+use crate::http;
+use crate::settings::global_user::GlobalUser;
+use crate::settings::toml::Target;
+
+const KEEP_ALIVE_INTERVAL: u64 = 60;
+
+pub struct Session;
+
+impl Session {
+    pub async fn run(
+        target: Target,
+        user: GlobalUser,
+        mut rx: Receiver<()>,
+    ) -> Result<(), failure::Error> {
+        let client = http::cf_v4_api_client_async(&user, HttpApiClientConfig::default())?;
+
+        let url = get_tunnel_url().await?;
+
+        let response = client
+            .request(&CreateTail {
+                account_identifier: &target.account_id,
+                script_name: &target.name,
+                params: CreateTailParams { url },
+            })
+            .await;
+
+        match response {
+            Ok(success) => {
+                eprintln!("Now prepared to stream logs.");
+
+                let tail_id = success.result.id;
+
+                // Loop indefinitely to send "heartbeat" to API and keep log streaming alive.
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+                let duration = Duration::from_millis(1000 * KEEP_ALIVE_INTERVAL);
+                let mut delay = time::delay_for(duration);
+
+                loop {
+                    match rx.try_recv() {
+                        Err(TryRecvError::Empty) => {
+                            if delay.is_elapsed() {
+                                let heartbeat_result =
+                                    send_heartbeat(&target, &client, &tail_id).await;
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = time::delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
+                    }
+                }
+            }
+            Err(e) => failure::bail!(http::format_error(e, None)),
+        }
+    }
+}
+
+async fn get_tunnel_url() -> Result<String, failure::Error> {
+    // regex for extracting url from cloudflared metrics port.
+    let url_regex = Regex::new("userHostname=\"(https://[a-z.-]+)\"").unwrap();
```

Comment:
TODO: what i'd really like to do is parse the output of the metrics endpoint as a stream (it's just prometheus metrics so there is a way to parse them) rather than using this regex but this is fine for RC.

Bonus points if we don't have to query this metrics endpoint at all but instead could get this as part of the initialization process but stdout is too wordy/subject to change; will try to work with tunnel team to figure out a good way to go about this.